### ISSUE TITLE
Implement Guetzli lossy compression for JPEG

### ISSUE THREAD

**Issue by fulldecent** [ImageOptim/ImageOptim#194]
Details here about a new Google publication

https://opensource.googleblog.com/2017/03/guetzli-new-open-source-jpeg-encoder.html?m=1

**Comment by pathawks** [comment]
https://github.com/google/guetzli/

**Comment by pornel** [comment]
I'm working on running a large benchmark (https://yardstick.pictures) on guetzli to see if there are any cases where it fails.

**Comment by HansVanEijsden** [comment]
Guetzli works great, wow. And if I run ImageOptim after Guetzli, it shaves off even more, about 5 - 7% (I keep the EXIF data).

**Comment by monouser7dig** [comment]
"Note that Guetzli is designed to work on high quality images (e.g. that haven't been already compressed with other JPEG encoders). While it will work on other images too, results will be poorer."
 from [https://github.com/google/guetzli#using](url)

would we have an option to put TIFF/PNG into imageoptim and output JPG?

thanks a lot for your effort, imageoptim is a great package so far!

**Comment by justo** [comment]
This is the best image optimization app out there, I would love to see this new tool added to it!

**Comment by pornel** [comment]
This is going to be delayed. Guetzli uses dangerously large amounts of memory. 

I tested it on a bunch of fairly small (<200KB) PNGs, and my 16GB system ran out of memory so hard, it became completely unresponsive and I had to force-reset the computer.

**Comment by mathiasbynens** [comment]
@pornel Did you mean JPEGs instead of PNGs in that last comment?

**Comment by pornel** [comment]
PNG as the source. JPEG to JPEG doesn't make sense for Guetzli.

**Comment by peterkovacs** [comment]
#196 is a quick first pass at integrating guetzli

**Comment by pornel** [comment]
Alpha released: https://imageoptim.com/changelog.html#v1.7.1

**Comment by pornel** [comment]
For now it's JPEG-only, so save JPEGs at 100% quality to use it. I'll enable it for PNG to JPEG conversion later.

It's runs one image at a time, because of huge memory requirements.

**Comment by timuric** [comment]
@pornel would be great to add quality adjustments when only Guetzli is enabled. For example when I use Guetzli from terminal with the -q 84 parameter, I am able to reduce the file size twice more compared to the results of the imageOptimizer

**Comment by pornel** [comment]
You can change the quality setting in the "Quality" tab in Preferences.

BTW: with Guetzli, the lower the quality setting, the closer it is to an old boring average JPEG encoder. 

And if you re-encode a Guetzli file, any further gain in the file size is only due to losing even more quality (it's like running 2x gives 2x file size savings, but 3x quality loss). So you can keep making the file smaller forever, but it's pointless, because at some point you'd get a better result even from Photoshop.

**Comment by timuric** [comment]
@pornel actually I have made a test in which I compared the results of Guetzli at the 84 quality and Imageoptim at 90, and I got better quality and lower size with Guetzli. 

Here are the results:

Original => 1.8mb 
http://d.pr/i/sRbU

ImageOtpim Q 90 => 486kb 
http://d.pr/i/k5pc

ImageOtpim Q 85 => 383kb ```
http://d.pr/i/uNIB

Guetzli Q 84 => 371kb
http://d.pr/i/Met1

Details comparison
http://d.pr/i/YBs1

Notice how Guetzli outperforms in terms of details even ```ImageOtpim Q 90``` and has lower size than ```ImageOtpim Q 84```

I have performed various tests on my image set, and Guetzli is all round winner for me (at the quality of 84)

**Comment by pornel** [comment]
By "ImageOptim" here do you mean ImageOptim with Guetzli disabled? So MozJPEG vs Guetzli?

**Comment by timuric** [comment]
Yes, sorry for not making that clear.

**Comment by timuric** [comment]
Agree, that MozJPEG performs better at lower compression rate, however Guetzli outperforms MozJPEG at the quality of 84 (size vs quality). Thus wouldn't be great to have control over Guetzli quality? Let say the slider could allow the user slide to minimum of 85 when only Guetzli​ is selected.

**Comment by pornel** [comment]
That's already the case (except Guetzli won't be run if you go below 85)

**Comment by timuric** [comment]
@pornel oh thats interesting. Is it expected behaviour that the slider is not movable if only Guetzli is selected?

**Comment by pornel** [comment]
Ahh, no. That's a bug :)

**Comment by fulldecent** [comment]
Google replaced this with https://github.com/google/jpegli
